lib/src/rules/avoid_returning_this_to_enable_a_fluent_interface.dart

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.avoid_returning_this_to_enable_a_fluent_interface;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r' ';
+
+const _details = r'''
+
+**DO** ...
+
+**BAD:**
+```
+
+```
+
+**GOOD:**
+```
+
+```
+
+''';
+
+bool _isExpressionFunctionBody(AstNode node) => node is ExpressionFunctionBody;
+
+bool _isReturnStatement(AstNode node) => node is ReturnStatement;
+
+bool _returnsThis(AstNode node) {
+  return (node as ReturnStatement).expression is ThisExpression;
+}
+
+class AvoidReturningThisToEnableAFluentInterface extends LintRule {
+  _Visitor _visitor;
+  AvoidReturningThisToEnableAFluentInterface()
+      : super(
+            name: 'avoid_returning_this_to_enable_a_fluent_interface',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitMethodDeclaration(MethodDeclaration node) {
+    if (node.returnType?.name?.name == 'void') {
```

Comment:
`TypeSystem.isSubtypeOf`

I don't know whether you have access to a type system, but that's what you need.